File: epi/IsValidParenthesization.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

public class IsValidParenthesization {
  static Map<Character, Character> sMap = new HashMap<>(){
    {
      this.put('(', ')');
      this.put('[', ']');
      this.put('{', '}');
    }
  };
  @EpiTest(testDataFile = "is_valid_parenthesization.tsv")
  public static boolean isWellFormed(String s) {
    /*
          ([]){()}
    stack {(
      T O(N) S O(N)
     */
    Deque<Character> stack = new ArrayDeque();
    for(char ch : s.toCharArray()) {
      if(sMap.containsKey(ch)) {
        stack.push(ch);
      } else {
        if(!stack.isEmpty() && sMap.get(stack.pop()) == ch) {

        } else {
          return false;
        }
      }
    }
    return stack.isEmpty();
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "IsValidParenthesization.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/IsValidParenthesization.kt
package epi.kt

import java.util.ArrayDeque

object IsValidParenthesization {
    fun isWellFormed(s: String): Boolean {
        var map = HashMap<Char, Char>()
        map[')'] = '('//round brackets
        map[']'] = '['//square brackets
        map['}'] = '{'//curly brackets
        var stack = ArrayDeque<Char>()
        for(i in s.indices) {
            var cur = s[i]
            if(!map.containsKey(cur)) {
                stack.push(cur)
            } else {
                if(stack.isEmpty() || stack.pop() != map[cur]) {
                    return false
                }
            }
        }

        return stack.isEmpty()
    }
}